feat(KL-105): add Product integration API

ohhamma opened the pull request:
## 📌 연관된 이슈
- **[KL-105/상품 통합 API 구현](https://ohhamma.atlassian.net/browse/KL-105)**

## 📝 작업 내용
- 상품 mvc 구현
- 상품 dto 구현 및 적용
- 상품 등록 API 구현
- 상품 상세 조회 API 구현
- 각 레이어의 테스트코드 작성

## 🌳 작업 브랜치명
- `KL-105/상품-통합-api-구현`

## 📸 스크린샷 (선택)

## 💬 리뷰 요구사항 (선택)

idealflower-k committed:
KL-107/feat: create product entity

idealflower-k committed:
KL-107/test: add product entity test code

idealflower-k committed:
Merge pull request #14 from taco-official/KL-107/삼품-엔티티-구현

feat(KL-107): 상품 엔티티 구현

idealflower-k committed:
KL-109/feat: create product repository

idealflower-k committed:
Merge pull request #15 from taco-official/KL-109/상품-repository-구현

feat(KL-109): 상품 repository 구현

idealflower-k committed:
KL-110/feat: create product dto

idealflower-k committed:
KL-110/feat: create product service interface

ohhamma committed:
Merge branch 'develop' into KL-105/상품-통합-api-구현

ohhamma committed:
Merge branch 'KL-105/상품-통합-api-구현' into KL-110/상품-service-구현

ohhamma committed:
KL-110/feat: modify eroduct field

- likeCount, price 필드 타입 Integer로 변경
- 변경된 Product에 맞게 테스트코드 수정

ohhamma committed:
KL-110/feat: modify ProductCreateRequestDto

- 선택 필드 지정
- price 필드 타입 Integer로 변경
- NotNull 메시지 한글로 변경

ohhamma committed:
KL-110/test: create ProductCreateRequestDtoTest

- 모든 필수 필드가 존재할 때 유효성 검사
- 개별 필수 필드가 null일 때 유효성 검사
- 여러 필수 필드가 null일 때 유효성 검사

ohhamma committed:
KL-110/feat: modify ProductUpdateRequestDto

- price 필드 타입 Integer로 수정

ohhamma committed:
KL-110/test: create ProductUpdateRequestDtoTest

- 모든 필드가 유효한 값일 때 유효성 검사
- 모든 필드가 null일 때 유효성 검사
- 일부 필드가 null일 때 유효성 검사

ohhamma committed:
KL-110/feat: modify ProductDetailResponseDto

- 불필요한 NotNull 삭제
- user 필드를 userId로 변경
- likeCount, price 필드 타입 변경
- 정적 팩토리 메서드 수정

ohhamma committed:
KL-110/test: create ProductDetailResponseDtoTest

- Product 객체로부터 ProductDetailResponseDto 생성 테스트
- ProductDetailResponseDto 생성자 테스트

ohhamma committed:
KL-110/feat: modify ProductSimpleResponseDto

- 불필요한 필드 제거
- 불필요한 NotNull 제거
- likeCount 필드 타입 변경
- 정적 팩토리 메서드 수정

ohhamma committed:
KL-110/test: create ProductSimpleResponseDtoTest

- Product 객체로부터 ProductSimpleResponseDto 생성 테스트
- 생성자를 통해 ProductSimpleResponseDto 생성 테스트

ohhamma committed:
KL-110/feat: add ProductNotFoundException

- 존재하지 않는 상품에 대한 ErrorCode 추가

ohhamma committed:
KL-110/feat: create DefaultConstants

- 자주 사용되는 기본값들을 저장해놓음

ohhamma committed:
KL-110/feat: modify Product

- NoArgsConstructor에 access 추가
- ColumnDefault에 DefaultConstants의 기본값으로 삽입
- likeCount와 createdAt 필드를 생성자에서 초기화하도록 수정

ohhamma committed:
KL-110/feat: create ProductNotFoundException

- 상품이 존재하지 않는 경우에 대한 예외 추가

ohhamma committed:
KL-110/feat: add findTestUser in UserUtil

- User과 관련된 util 메소드들을 담는 UserUtil 생성
- 임시 테스트용 유저를 반환하는 findTestUser 메소드 작성

ohhamma committed:
KL-110/feat: create product and get product info

- ProductService 인터페이스 삭제 및 클래스로 변환
- Product 생성
- id로 Product의 상세 정보 조회

ohhamma committed:
KL-110/test: create ProductServiceTest

- ID로 상품 정보를 조회할 때, 상품이 존재하면 ProductDetailResponseDto 반환
- ID로 상품 정보를 조회할 때, 상품이 존재하지 않으면 ProductNotFoundException 발생
- 상품을 생성할 때, 정상적으로 ProductDetailResponseDto 반환

ohhamma committed:
Merge pull request #18 from taco-official/KL-110/상품-service-구현

feat(KL-110) : create Product service

ohhamma committed:
KL-111/chore: add dummy Product data

- 테스트용 상품 더미 데이터 생성

ohhamma committed:
KL-111/feat: create ProductController

- 상품 Controller 작성
- 상품 상세 정보를 조회하는 API 추가
- 상품을 등록하는 API 추가

ohhamma committed:
KL-111/test: create ProductControllerTest

- 상품 상세 조회 API 테스트
- 상품 등록 API 테스트

ohhamma committed:
KL-111/test: create ProductIntegrationTest

- 상품 상세 조회 API 테스트
- 상품 등록 API 테스트

ohhamma committed:
Merge pull request #22 from taco-official/KL-111/상품-controller-구현

feat(KL-111): add Product controller

sayoonnn commented:
LGTM!

min3m commented on src/main/java/taco/klkl/domain/product/domain/Product.java:
```diff
@@ -0,0 +1,111 @@
+@Getter
+@NoArgsConstructor
+@Entity(name = "product")
+@DynamicInsert
```
DynamicInsert가 어떤 경우 사용된다고 했었죠??

ohhamma replied:
`@DynamicInsert`는 db에 레코드를 집어넣는 과정에서 null이 아닌 필드만 동적으로 골라내서 `INSERT` 쿼리문을 실행해주는 어노테이션입니다! 밑에 있는 `@ColumnDefault`는 db 스키마를 생성하는 과정에서 필드값의 기본값을 지정해주는 어노테이션으로 조금 다른 의미로 해석하시면 될 것 같아요.

min3m commented on src/main/java/taco/klkl/domain/product/domain/Product.java:
```diff
@@ -0,0 +1,111 @@
+	@Column(name = "currency_id", nullable = false)
+	private Long currencyId;
+
+	@PrePersist
```
PrePersist는 어떤 경우 작동되고 사용되는건가요?

ohhamma replied:
`@PrePersist` 어노테이션이 붙은 메소드는 데이터가 db에 영속적으로 저장되기 직전에 실행됩니다. 그래서 보통 유효성 검사나 초기화 작업이 복잡해지는 경우에 사용한다고 하네요. 또, `@PrePersist`를 사용하면 데이터의 무결성을 db의 레벨에 의존하는 것이 아닌, 어플리케이션 레벨에서도 지킬 수 있다는 장점이 있어요.

min3m commented on src/main/java/taco/klkl/domain/product/dto/response/ProductSimpleResponseDto.java:
```diff
@@ -0,0 +1,38 @@
+ * @param subcategoryId
+ * @param currencyId
+ */
+public record ProductSimpleResponseDto(
```
ResponseDto에서 NotNull 어노테이션을 붙여준 이유가 무엇인가요?

ohhamma replied:
ResponseDto가 아니라 **RequestDto**에서 주로 `@NotNull`을 붙여줍니다! 응답으로 보낼 때 사용하는 dto에는 서버에서 값을 직접 넣어줄 거기 때문에 null 확인을 할 필요가 없습니다. 반면, 클라이언트로부터 들어온 요청을 담는 dto에서는 필드가 모두 잘 들어왔는지 확인하기 위해서 `@NotNull`을 사용합니다.

min3m replied:
엇 그렇죠? ㅋㅋㅋ 아니 ResponseDto에도 Notnull이 들어 있던걸 봤는데... 잘못봤나봅니다... ㅋㅋㅋ

min3m commented on src/test/java/taco/klkl/domain/product/dto/request/ProductCreateRequestDtoTest.java:
```diff
@@ -0,0 +1,182 @@
+	private Validator validator;
+
+	@BeforeEach
+	void setUp() {
```
여기서 셋업 해주는 ValidatorFactory라는 객체는 무엇인가요? 
또 어떤 식으로 Validate 동작이 진행되나요?

ohhamma replied:
위에서 언급했둣이 필수적으로 들어와야하는 필드에 대해서 `@NotNull` 어노테이션을 붙이는데, 이 부분이 잘 작동하는지 확인하는 과정에서 Validator를 사용합니다. 해당 어노테이션을 붙인 필드들에 null값이 들어오면 누락된 필드의 개수만큼 violation이 발생하고, 각 violation이 앞에서 설정한 에러 메시지와 일치하는지 확인하는 작업을 거치는 형식으로 유효성 검사를 합니다.

따라서 `@NotNull`이 붙지 않는 ResponseDto의 테스트코드에서는 Validator을 사용하지 않는다는 사실을 알 수 있습니다 !

ohhamma replied:
추가로, `ValidatorFactory`는 이러한 유효성 검사를 실행해주는 `Validator`를 만들어주는 객체라고 합니다. 말그대로 팩토리, 공장인 셈이에요 !

```java
// ValidatorFactory 생성
ValidatorFactory factory = Validation.buildDefaultValidatorFactory();

// Validator 인스턴스 얻기
Validator validator = factory.getValidator();
```

ohhamma replied:
`validate()`의 자세한 작동 방식은 아래와 같습니다.

1. 리플렉션을 사용하여 객체의 필드, 메서드, 클래스 레벨의 제약 조건(annotations)을 검사합니다.
2. 각 제약 조건에 대해 해당하는 `ConstraintValidator`를 실행합니다.
3. 제약 조건을 위반한 경우, `ConstraintViolation` 객체를 생성하여 결과 `Set`에 추가합니다.
4. 모든 제약 조건 검사가 완료되면, `ConstraintViolation` 객체들의 `Set`을 반환합니다.

```java
// 유효성 검사 실행
Set<ConstraintViolation<ProductCreateRequestDto>> violations = validator.validate(dto);
```

min3m commented on src/test/java/taco/klkl/domain/product/dto/request/ProductCreateRequestDtoTest.java:
```diff
@@ -0,0 +1,182 @@
+
+	@Test
+	@DisplayName("여러 필수 필드가 null일 때 유효성 검사 실패")
+	void testProductCreateRequestDtoWithMultipleNullFields() {
```
이렇게 여러개의 항목이 null일경우 모든 경우에 대해서 예외를 발생시키나요?

ohhamma replied:
정확히는, 누락된 필드의 개수만큼 violation이 발생해야 합니다. 이 코드는 여러 필수 필드가 null인 경우에 `@NotNull`에 따라 누락된 필드의 개수와 유효성 검사 시 발생한 violation의 개수가 일치하는지 여부를 확인하기 위해 만들어둔 테스트케이스입니다. 따라서, 누락된 필드의 개수가 3개인 dto로 테스트코드를 실행했을 때에는 발생한 violation 개수가 3개인지 확인하면 되겠죠.

min3m commented on src/test/java/taco/klkl/domain/product/dto/response/ProductDetailResponseDtoTest.java:
```diff
@@ -0,0 +1,109 @@
+
+	@Test
+	@DisplayName("ProductDetailResponseDto 생성자 테스트")
+	void testConstructor() {
```
Product를 이용해서 생성하는 경우만 test하면 되지 않나요? 
생성자를 통해 ResponseDto를 생성하는 테스트를 만드신 이유가 있을 까요?

ohhamma replied:
보통 생성자를 테스트하는 경우는 다음과 같다고 합니다.

> 1. **복잡한 로직이 있는 경우**: 생성자 내부에 복잡한 로직이 포함되어 있거나, 입력 값에 따라 다른 동작을 수행하는 경우.
> 2. **유효성 검사**: 생성자 내부에서 입력 값에 대한 유효성 검사를 수행하는 경우.
> 3. **상태 초기화**: 생성자가 객체의 상태를 초기화하는데 중요한 역할을 하는 경우.

그러나 현재 테스트를 진행하고 있는 `ProductDetailResponseDto`에서는 위의 3가지 경우에 해당되지 않네요. 다만, 다음과 같은 이유로 생성자에 테스트를 해도 괜찮다고 합니다.

> 1. **안정성**: 생성자가 예상대로 객체를 초기화하는지 확인할 수 있습니다.
> 2. **변경 감지**: 생성자에 변경이 생길 경우, 테스트가 실패하여 이를 감지할 수 있습니다.
> 3. **일관성 유지**: 다른 테스트와의 일관성을 유지할 수 있습니다.

만든 이유는 따로 없으나, 앞으로 일관성에 맞게 생성자 테스트를 모두 지우거나 모두 생성하거나 할 것 같습니다 !

min3m replied:
흠 Response 객체에서 생성자나 팩토리 매서드가 잘 작동하는지 테스트는 한번 고민을 해봐야될거 같아요. 
특히나 팩토리 매서드는 차치하고 기본생성자는 굳이?라는 생각이 드네요!

min3m commented on src/main/java/taco/klkl/domain/product/service/ProductService.java:
```diff
@@ -1,22 +1,49 @@
+		return ProductDetailResponseDto.from(product);
+	}
+
+	public ProductDetailResponseDto createProduct(final ProductCreateRequestDto productDto) {
```
createProduct와 createProducEntity함수를 분리한 이유가 있을까요?

ohhamma replied:
`Product` 엔티티를 생성하기 위해서는 현재 코드 기준 `User` 객체가 따로 필요합니다. 그러나 `createProduct`에서는 Product 객체를 repository에 save하고, responseDto를 반환하는 로직만 들어가는 것이 맞다고 생각했습니다. 또 이후 `City`, `Subcategory`, `Currency` 등 `Product` 엔티티를 생성하는 과정에서 점점 더 많은 로직이 생겨날 것을 우려하여 두 로직을 다른 메소드로 분리했습니다. + 그리고 S3 로직도 들어갈 예정입니다 ...

min3m commented on src/main/java/taco/klkl/domain/product/controller/ProductController.java:
```diff
@@ -0,0 +1,40 @@
+
+	@PostMapping
+	@Operation(summary = "상품 등록", description = "상품을 등록합니다.")
+	public ResponseEntity<ProductDetailResponseDto> createProduct(@RequestBody ProductCreateRequestDto createRequest) {
```
RequestDto를 RequestBody 어노테이션으로 매핑시킨건가요?

ohhamma replied:
네 맞습니다! `@RequestBody`는 HTTP 요청 본문에 포함된 데이터를 자바 객체로 변환하는 데 사용되는 어노테이션입니다.

idealflower-k commented on src/main/java/taco/klkl/global/common/constants/DefaultConstants.java:
```diff
@@ -0,0 +1,12 @@
+package taco.klkl.global.common.constants;
+
+public class DefaultConstants {
```
Enum이나 환경변수가 아닌 클래스로 처리한 이유가 궁금합니다!

min3m replied:
엇 저도 이거 궁금했는데

ohhamma replied:
Enum을 사용하기 위해서는 모든 상수들이 동일한 타입의 값을 가져야 하는데, 이후에 `String` 타입 이외에도 int 등 다른 타입의 상수들도 들어올 것을 우려하여 클래스로 처리했습니다. 근데 나중에 구현하고 보니 상수들이 모두 같은 형태를 지닌 경우에는 Enum으로 관리하는 것이 더 좋은 패턴이겠네요.

idealflower-k commented on src/main/java/taco/klkl/global/error/exception/ErrorCode.java:
```diff
@@ -16,6 +16,7 @@ public enum ErrorCode {
 	// User
 
 	// Product
+	PRODUCT_NOT_FOUND(HttpStatus.NOT_FOUND, "C200", "존재하지 않는 상품입니다."),
```
020이 아니라 200부터 시작이었나요?

ohhamma replied:
헉 수정하겠습니다 !

idealflower-k commented on src/main/java/taco/klkl/global/util/UserUtil.java:
```diff
@@ -0,0 +1,19 @@
+
+@Component
+@RequiredArgsConstructor
+public class UserUtil {
```
UserService가 import되어있는데 sevice가 아닌 repository를 이용한 이유가 궁금합니다!

ohhamma replied:
`UserService`가 import된 이유는 단순히 .. 상수값을 가져오기 위함이었습니다. 아예 이 부분을 constants로 빼는게 더 좋겠네요. 실제로는 `UserRespository`를 사용하여 테스트용 유저를 가져오는 게 맞습니다 !

min3m reviewed (dismissed):
Test 코드 정말 빡빡하게 작성하느라 고생많으셨습니다... 사실상 테스트코드가 이정도면 본채 이지 않을까 싶네요.. ㅋㅋㅋ

idealflower-k commented on src/test/java/taco/klkl/domain/product/domain/ProductTest.java:
```diff
@@ -0,0 +1,108 @@
+import taco.klkl.domain.user.domain.User;
+
+class ProductTest {
+	private static final Integer DEFAULT_PRICE = 0;
```
DefaultConstants 클래스에 정의된 값들을 가져와 비교하는게 더 좋은 테스트가 될거 같습니다!

ohhamma replied:
저도 동의합니다! 도메인별 constants들을 만들어서 따로 관리하도록 수정하겠습니다.

ohhamma committed:
KL-105/chore: change ProductNotFoundException code

- 유효하지 않은 상품인 경우의 예외에 대한 코드 "C020"으로 변경

ohhamma committed:
KL-105/chore: add ProductConstants

- Product 관련 상수값을 모아놓은 ProductConstants 작성
  - 필드 기본값 (가격, 좋아요수, 주소)
- Product 관련 코드에 모두 적용

ohhamma committed:
KL-105/chore: add UserConstants

- User 관련 상수값들을 모아둔 UserConstants 작성
  - 테스트용 유저
  - 필드 기본값 (프로필, 총 좋아요수)
- User 관련 코드에 모두 적용

min3m reviewed (approved):
Test 코드 정말 빡빡하게 작성하느라 고생많으셨습니다... 사실상 테스트코드가 이정도면 본채 이지 않을까 싶네요.. ㅋㅋㅋ

idealflower-k reviewed (approved):
양이 많은데 작성하시느라 고생하셨습니다!
상품과 유저까지 Default 클래스로 관리하는 걸로 수정된거 확인했습니다!
고생하셨습니다! :)
